File: src-tauri/src/models/recording.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

// ── Upload row (maps to recording_uploads table) ──────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
#[serde(rename_all = "camelCase")]
pub struct RecordingUpload {
    pub recording_id: Uuid,
    pub platform: String,
    pub state: String,
    pub progress_bytes: i64,
    pub total_bytes: i64,
    pub visibility: String,
    pub video_id: Option<String>,
    pub video_url: Option<String>,
    pub error: Option<String>,
    pub started_at: Option<DateTime<Utc>>,
    pub completed_at: Option<DateTime<Utc>>,
    pub updated_at: DateTime<Utc>,
}

// ── Main recording row ────────────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
#[serde(rename_all(serialize = "camelCase", deserialize = "snake_case"))]
pub struct Recording {
    pub id: Uuid,
    pub event_id: Uuid,
    pub file_path: String,
    pub file_name: String,
    pub file_size: i64,
    pub duration_seconds: f64,
    pub detected_at: DateTime<Utc>,
    pub whitelisted: bool,
    pub uploaded: bool,
    pub uploaded_at: Option<DateTime<Utc>>,
    pub video_id: Option<String>,
    pub video_url: Option<String>,
    pub custom_title: Option<String>,
    pub uploadable: bool,
    pub custom_description: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    /// Populated manually after a JOIN — not a DB column on recordings.
    #[sqlx(skip)]
    #[serde(default)]
    pub uploads: Vec<RecordingUpload>,
}

// ── Request bodies ────────────────────────────────────────────────────────────

/// Received from frontend — stays snake_case to match JSON body
#[derive(Debug, Deserialize)]
pub struct CreateRecording {
    pub file_path: String,
    pub file_name: String,
    pub file_size: Option<i64>,
    pub duration_seconds: Option<f64>,
    pub custom_title: Option<String>,
    pub custom_description: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct FlagUploadRequest {
    pub recordings: Vec<FlagUploadItem>,
}

#[derive(Debug, Deserialize)]
pub struct FlagUploadItem {
    pub recording_id: Uuid,
    pub custom_title: Option<String>,
    pub custom_description: Option<String>,
    pub youtube_visibility: Option<String>,
    pub facebook_visibility: Option<String>,
    pub platforms: Vec<String>,
}

File: src/lib/schemas/recording.ts
import { z } from 'zod';

export const RecordingUploadSchema = z.object({
  recordingId: z.string().uuid(),
  platform: z.string(),
  state: z.enum(['pending', 'uploading', 'paused', 'completed', 'failed']),
  progressBytes: z.number().int(),
  totalBytes: z.number().int(),
  visibility: z.string(),
  videoId: z.string().nullish(),
  videoUrl: z.string().nullish(),
  error: z.string().nullish(),
  startedAt: z.string().nullish(),
  completedAt: z.string().nullish(),
  updatedAt: z.string(),
});

export type RecordingUpload = z.infer<typeof RecordingUploadSchema>;

export const RecordingSchema = z.object({
  id: z.string().uuid(),
  eventId: z.string().uuid(),
  filePath: z.string(),
  fileName: z.string(),
  fileSize: z.number().int().nonnegative(),
  durationSeconds: z.number().nonnegative(),
  detectedAt: z.string(),
  whitelisted: z.boolean(),
  uploaded: z.boolean(),
  uploadedAt: z.string().nullish(),
  videoId: z.string().nullish(),
  videoUrl: z.string().nullish(),
  customTitle: z.string().nullish(),
  uploadable: z.boolean().default(false),
  customDescription: z.string().nullish(),
  uploads: z.array(RecordingUploadSchema).default([]),
  createdAt: z.string(),
  updatedAt: z.string(),
});

export type Recording = z.infer<typeof RecordingSchema>;

export const CreateRecordingPayloadSchema = z.object({
  file_path: z.string().min(1),
  file_name: z.string().min(1),
  file_size: z.number().int().nonnegative().optional(),
  duration_seconds: z.number().nonnegative().optional(),
  custom_title: z.string().optional(),
  custom_description: z.string().optional(),
});

export type CreateRecordingPayload = z.infer<typeof CreateRecordingPayloadSchema>;

export const FlagUploadItemSchema = z.object({
  recording_id: z.string().uuid(),
  custom_title: z.string().optional(),
  custom_description: z.string().optional(),
  youtube_visibility: z.enum(['private', 'unlisted', 'public']).optional(),
  facebook_visibility: z.enum(['ONLY_ME', 'FRIENDS', 'EVERYONE']).optional(),
  platforms: z.array(z.string()),
});

export type FlagUploadItem = z.infer<typeof FlagUploadItemSchema>;

export const FlagUploadRequestSchema = z.object({
  recordings: z.array(FlagUploadItemSchema),
});

export type FlagUploadRequest = z.infer<typeof FlagUploadRequestSchema>;

export const RecordingWithEventSchema = RecordingSchema.extend({
  eventTitle: z.string(),
});
export type RecordingWithEvent = z.infer<typeof RecordingWithEventSchema>;
